### PR TITLE
Command Api v2 based on SpongeAPI's command api

### PR DESCRIPTION
**Pull request by BoogieMonster1O1**
The previous two command APIs didn't really work out well. Commands v0 provides helper classes to deal with Vanilla commands, which don't have any special parsing capabilities and hence are a pain to work with when dealing with commands that have arguments. While commands v1 does solve that problem by using brigadier, it currently does not work and ramidzkh hasn't gotten themself to fix it. 

Command API v2 uses SpongeAPI's command API for a number of reasons
- Sponge's command API is lesser verbose than brigadier (no `StringReader`s involved)
- Sponge's command API can directly interact with vanilla stuff without having to implement API classes (such as `Message`)
- Commands v1 does not provide any hooks or classes for commonly used argument types
- Brigadier and vanilla commands do not support permissions out of the box
- Sponge's command API has more documentation
- Sponge's command API is made to work in a non-brigadier environment (1.12)
- Sponge's command API isn't made by ojank

The new command api includes highly customized versions of SpongeAPI's `org.spongepowered.api.command` classes in the `net.fabricmc.fabric.api.command.v2.lib.sponge` package. License headers are preserved. 

Features
- [x] Api
- [x] Implementation
- [x] Testmod

Requires #47

### CODE REVIEW

**Comment by hYdos** [comment]
i find it a better solution if we just fix brigadier. i was talking to 15 and you just gotta change the ramp task a bit.

**Comment by BoogieMonster1O1** [comment]
> i find it a better solution if we just fix brigadier. i was talking to 15 and you just gotta change the ramp task a bit.

who's 15?
#47 fixes the building and mixin issues but doesn't fix the api itself
v1 also doesn't provide argument types and message hooks
this doesn't even need those as spongeapi has them

**Comment by hYdos** [comment]
i509VCB

**Comment by BoogieMonster1O1** [comment]
and suggestions in custom argument types are also not simple in brigadier
1.13+ has an entire class for suggestions

**Comment by hYdos** [comment]
fabric api is meant to be **Hooks and Essential** API. not add every other library to handle stuff for us API. it also causes dependency hell.

**Comment by BoogieMonster1O1** [comment]
> fabric api is meant to be **Hooks and Essential** API. not add every other library to handle stuff for us API. it also causes dependency hell.

that's what happened with commands v1.
 and that wasn't something like loader and gson where parts of it are repackaged and modified. 
that was an entirely new unmodified dependency that had to be shaded
broke commands v0
broke viafabric :(

**Comment by hYdos** [comment]
this will not get merged. its brigadier or generic 1.8.9 commands. do **not** reopen this boogie.

**Comment by BoogieMonster1O1** [comment]
it compiles 🥳 
ive excluded the files in `lib/sponge` from licenser because those have spongeapi headers

**Comment by BoogieMonster1O1** [comment]
finished the testmod. haven't tested it yet
@hYdos review

**Comment by BoogieMonster1O1** [comment]
The testmod works
https://imgur.com/E0c8DF7

**Comment by BoogieMonster1O1** [comment]
@shedaniel review

File: build.gradle [hunk 1]
```diff
@@ -121,8 +121,10 @@ allprojects {
 		}
 
 		license {
-			header rootProject.file('HEADER')
+			header = rootProject.file('HEADER')
```
Comment (i509VCB) [inline build.gradle:124]: lmao unneeded change

File: build.gradle [hunk 2]
```diff
@@ -121,8 +121,10 @@ allprojects {
...
 			include '**/*.java'
+			exclude '**/lib/sponge/*.java'
+			exclude '**/lib/sponge/**/*.java'
```
Comment (i509VCB) [inline build.gradle:127]: you know you can auto apply the sponge license headers in here to those files...

File: fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java [hunk 3]
```diff
@@ -17,14 +17,76 @@
 package net.fabricmc.fabric.api.util;
 
 public enum TriState {
-	FALSE,
-	DEFAULT,
-	TRUE;
+	TRUE(true) {
```
Comment (i509VCB) [inline fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java:20]: You know we wrote some tristate extensions that are now in upstream...

File: fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java [hunk 4]
```diff
@@ -17,14 +17,76 @@
...
+		}
+
+		@Override
+		public TriState or(TriState other) {
+			return other;
+		}
+
+		@Override
+		public TriState negate() {
+			return this;
```
Comment (i509VCB) [inline fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java:65]: Undocumented behavior

File: fabric-armor-api-v1/src/testmod/java/net/fabricmc/fabric/test/armor/ArmorTest.java [hunk 5]
```diff
@@ -1,67 +0,0 @@
...
-import net.minecraft.item.Items;
-import net.minecraft.util.Identifier;
-
-import net.fabricmc.api.ModInitializer;
-import net.fabricmc.fabric.api.armor.v1.ArmorMaterial;
-import net.fabricmc.fabric.api.content.registry.v1.ItemRegistry;
-import net.fabricmc.fabric.impl.armor.EquipmentSlot;
-import net.fabricmc.fabric.impl.armor.FabricArmorItem;
-
-public class ArmorTest implements ModInitializer {
```
Comment (i509VCB) [inline fabric-armor-api-v1/src/testmod/java/net/fabricmc/fabric/test/armor/ArmorTest.java:32]: :thinking:

**Review by CoolMineman — CHANGES_REQUESTED** [review]
6.5/10 too much code
concern

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/Location.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.command.v2;
+
+import net.minecraft.util.math.BlockPos;
+
+public class Location<T> {
```
Comment (i509VCB) [inline fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/Location.java:21]: Why is this even needed? Couldn't the world and pos be passed as seperate parameters

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/CommandResult.java [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+		this.affectedItems = Optional.ofNullable(affectedItems);
+		this.queryResult = Optional.ofNullable(queryResult);
+	}
+
+	/**
+	 * Gets the success count of the command.
+	 *
+	 * @return The success count of the command
+	 */
+	public Optional<Integer> getSuccessCount() {
```
Comment (i509VCB) [inline fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/CommandResult.java:151]: What ever happened to `OptionalInt`

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/ArgUtils.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+import javax.annotation.Nullable;
+
+import net.minecraft.text.Text;
+import net.minecraft.text.TranslatableText;
+
+/**
+ * Internal utility methods.
+ */
+class ArgUtils {
```
Comment (i509VCB) [inline fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/ArgUtils.java:36]: Is this not what the impl package is for?

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/GenericArguments.java [hunk 9]
```diff
@@ -0,0 +1,2184 @@
...
+	 * under their provided keys in the CommandContext.
+	 *
+	 * @param element The element to repeat
+	 * @return the element to match the input
+	 */
+	public static CommandElement allOf(CommandElement element) {
+		return new AllOfCommandElement(element);
+	}
+
+	private static class AllOfCommandElement extends CommandElement {
```
Comment (i509VCB) [inline fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/GenericArguments.java:819]: This thing is an inner class nightmare, move all the impl classes to impl package

**Review by i509VCB — CHANGES_REQUESTED** [review]
@hYdos you can nuke this at will

File: build.gradle (hunk 2 above)
Comment (BoogieMonster1O1) [reply build.gradle:127]: already did

File: fabric-armor-api-v1/src/testmod/java/net/fabricmc/fabric/test/armor/ArmorTest.java (hunk 5 above)
Comment (BoogieMonster1O1) [reply fabric-armor-api-v1/src/testmod/java/net/fabricmc/fabric/test/armor/ArmorTest.java:32]: it doesnt work

File: fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java (hunk 3 above)
Comment (BoogieMonster1O1) [reply fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java:20]: sure k will diff

File: build.gradle (hunk 1 above)
Comment (BoogieMonster1O1) [reply build.gradle:124]: i actually nuked thwe whole thing a couple of times
ended up with `=`

File: build.gradle (hunk 2 above)
Comment (BoogieMonster1O1) [reply build.gradle:127]: actually sponge license headers were already applied so i didnt care

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/CommandResult.java (hunk 7 above)
Comment (BoogieMonster1O1) [reply fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/CommandResult.java:151]: blame api 7

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/ArgUtils.java (hunk 8 above)
Comment (BoogieMonster1O1) [reply fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/ArgUtils.java:36]: yes but actually no

Comment (BoogieMonster1O1) [reply fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/ArgUtils.java:36]: fine ill move it

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/GenericArguments.java (hunk 9 above)
Comment (BoogieMonster1O1) [reply fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/GenericArguments.java:819]: blame api 7
inner classes good and best

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/Location.java (hunk 6 above)
Comment (BoogieMonster1O1) [reply fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/Location.java:21]: yes but Location is more spongy

Comment (BoogieMonster1O1) [reply fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/Location.java:21]: fine ill crob this too

**Comment by hYdos** [comment]
image
the final nail in the coffin

**Comment by NeRdTheNed** [comment]
1A28657A-10D9-4F62-B7C2-7D52F77DB872

@hYdos Looks like a tie to me :tiny_potato:

File: fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java (hunk 4 above)
Comment (BoogieMonster1O1) [reply fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java:72]: ficksed

**Review by LePichu — APPROVED** [review]
this good and best and i approve because boogie dumb :tiny_potati:

**Comment by BoogieMonster1O1** [comment]
I've split up more parts of the api and removed the inner classes in `GenericArguments`

File: fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/CommandResult.java (hunk 7 above)
Comment (BoogieMonster1O1) [reply fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/CommandResult.java:151]: the values are nullable

**Review by hYdos — APPROVED** [review]
ok but i still dislike sponge. now get @i509VCB to re-review and approve

**Review by ramidzkh — APPROVED** [review]
the voice from the closet says it'll shoot me if i dont approve

**Comment by BoogieMonster1O1** [comment]
hydos will never merge this

**Comment by BoogieMonster1O1** [comment]
yes it shalt return soon

**Comment by BoogieMonster1O1** [comment]
wow this has been bikeshedded for around four months now

**Review by BluCobalt — APPROVED** [review]
image

**Comment by BluCobalt** [comment]
Before this gets merged, can you verify that this doesn't crash?

**Comment by BoogieMonster1O1** [comment]
> Before this gets merged, can you verify that this doesn't crash?

i sent a video in showcase

**Comment by BluCobalt** [comment]
would it be possible for you to send a link?

**Comment by BluCobalt** [comment]
image
proof

**Pull request merged**